src/coreclr/vm/nativelibrary.cpp

```diff
@@ -712,10 +712,18 @@ namespace
                 }
             }
 
-            hmod = LocalLoadLibraryHelper(currLibNameVariation, dllImportSearchPathFlags, pErrorTracker);
-            if (hmod != NULL)
+            // Internally, search path flags and whether or not to search the assembly directory are
+            // tracked separately. However, on the API level, DllImportSearchPath represents them both.
+            // When a user specifies DllImportSearchPath.AssemblyDirectory, searchAssemblyDirectory is
+            // true, dllImportSearchPathFlags is 0, and the desired logic is to only search the assembly
+            // directory (handled above), so we avoid doing any additional load search in that case.
+            if (!userSpecifiedSearchFlags || !searchAssemblyDirectory || dllImportSearchPathFlags != 0)
```

Comment:
It's really ugly that `searchAssemblyDirectoy = TRUE, dllImportSearchPathFlags = 0` is the same value we use for the "default" case, but I guess that's just what it is (and why this bug wasn't found until now).